aws/src/main/java/org/apache/iceberg/aws/s3/S3OutputStream.java

```diff
@@ -74,13 +78,16 @@
   private final AwsProperties awsProperties;
 
   private CountingOutputStream stream;
-  private final List<File> stagingFiles = Lists.newArrayList();
+  private final List<FileAndDigest> stagingFiles = Lists.newArrayList();
   private final File stagingDirectory;
   private File currentStagingFile;
   private String multipartUploadId;
   private final Map<File, CompletableFuture<CompletedPart>> multiPartMap = Maps.newHashMap();
   private final int multiPartSize;
   private final int multiPartThresholdSize;
+  private final boolean isEtagCheckEnabled;
+  private final MessageDigest completeMessageDigest;
+  private MessageDigest currentPartMessageDigest;
```

Comment:
Rather than computing all of the digests ourselves, it might be cleaner to wrap the streams in `java.security.DigestOutputStream` and delegate to that in order to calculate the etag.  I'm not sure if that approach can be used for the complete message (since the parts are uploaded independently), but I also question whether we really need to validate the full checksum if we get all of the parts verified.

Reply:
If we do want to calculate the full stream, it looks like it may be possible to wrap it twice: once for the individual part streams where we reset the digest every time, and a second wrapped stream where we reuse the digest across each file.

Reply:
Thanks @danielcweeks for the review. We don't do full checksum if we are doing multi-part upload, we only do full checksum if it is a single put request. If I am understanding correctly wrapping the stream with `DigestOutputStream` will make us handle counts explicitly, which will beat the purpose of this suggestion to simplify the changes. Please correct me if I am missing something here.